### ISSUE TITLE
Thread pool management

### ISSUE THREAD

**Issue by jdegoes** [scalaz/scalaz-zio#248]
Thread pool management is an unsolved problem in the ecosystem, with some initial attempts provided by [linebacker](https://christopherdavenport.github.io/linebacker/) and `cats-effect` `evalOn`, which shifts to and from an `ExecutionContext` under `bracket`.

A solution must contend with the following competing goals:

1. Aggressively adding threads without bound will decrease throughput.
2. Failure to add threads will result in task starvation when `IO` is used with blocking IO (e.g. JDBC), resulting in a system-wide failure to make progress.
3. Some third-party libraries wish work to be performed on their own thread pools.

`IO` has no guarantee that a computation will run on the same thread or even in the same thread pool. Due to the `async` constructors, it's impossible to make such a guarantee. Yet, the failure to have such a guarantee makes reasoning about thread pool management difficult.

Further, the `async` constructors add additional problems in Javascript code (via `Scala.js`), because some Javascript code must happen inside the current tick of the clock (for example, canceling an event bubble must be done synchronously). 

In order to solve these problems and provide clear semantics, I propose adding a new `SyncIO` type. This type will not have any `async` constructors, will only ever execute on a single thread, and will have no other capability to "jump" to another thread pool.

The following capabilities can be provided:

1. The ability to run a `SyncIO` on a thread pool. e.g. `def run(ec: ExecutionContext): IO[Nothing, Fiber[E, A]]`. 
2. The ability to run a `SyncIO` on the "current" thread from which it is invoked; e.g. `def run: IO[E, A]`.
3. The ability to interrupt a `SyncIO` task via both normal interruption (inside the interpreter, at every operation), but also `Thread.interrupt`, without any additional performance cost.

Then `SyncIO` should be encouraged for all blocking IO, for all Javascript callbacks, and for any code that is destined to run inside a third-party library's thread pool. Because this IO has no capability to jump threads, it is easy to reason about its thread behavior; with safe and typed interface points for `IO`, a user is carroted toward safe integration of synchronous IO into programs with mixed synchronicity. 

As added advantages of this approach:

1. `Thread.interrupt` can interrupt blocking IO on evaluating `SyncIO`, allowing the results of long-running IO operations, such as JDBC, to be interrupted and cleaned up safely. Incorporating this into `IO` leads to performance problems that are very difficult to solve.
2. The primacy of `IO` is secured, as there is no way to "run" a `SyncIO` via `unsafeRun`, and no way to convert a `SyncIO` into an `IO` (which would have poorly-defined semantics given the intended use case of `SyncIO`). The only way to "run" a `SyncIO` is to submit it to a definite and non-implicit `ExecutionContext`, which means the user is forced to choose exactly where the operation should be run.
3. The only means for an `IO` program to interact with a `SyncIO` program is by means of the `Fiber`. Fiber has limited composition semantics (it's result can be joined in `IO`, interrupted, or observed); none of these means of composition interfere or change the guarantees. Users may safely interact with `SyncIO` code without having to think or be careful.
4. `IO` has increased capability to be extremely aggressive about pre-emptive yielding, because `SyncIO` can be used for all other use cases that do not permit pre-emption. 

/cc @regiskuckaertz 

See also: #241 #75

**Comment by pshirshov** [comment]
I have to add my 5 cents:

1. Scheduling problem is a fundamental problem
2. You can't do a lot in case you don't know the shape of your dependency graph ahead of time. So you have to use heuristics
3. Each heuristic can be spoiled. Ideally I would prefer to have a basic zio interface which can be implemented in completely sync manner, without any scheduling.

**Comment by jdegoes** [comment]
There's no need to do scheduling as part of this proposal. Though if there is no scheduling then an unbounded thread pool is the only way to guarantee software correctness (otherwise unintended starvation can lead to deadlocks).

**Comment by pshirshov** [comment]
Just a note on terminology: when we submit tasks into an unbounded threadpool it's a "scheduling" too, just one of the simplest form. And yup, in case the pool is limited there is always a risk of exhaustion. It would be kinda incorrect to call it "deadlock" - by definition of deadlock. "starvation" term is widely used to describe situation when you have not enough threads to finish computation, though some researchers use it to describe a situation when workers are underloaded and throughput drops drastically because of bad scheduling so I prefer to say "exhaustion" instead.

 Though its hard to schedule properly when you have no global knowledge on your _project network_ - a graph of dependencies between your operations. 

Especially in a situation when we cannot save current computation context blocked by an op and serve another op (do "context switch" or "continuation")

In fact it's proven to be impossible to invent an ideal scheduling strategy.

So, the from purely theorethical point of view you may:

- choose a heuristic (easy but easy to spoil)
- employ continuations (very hard)

At the same time in many cases you don't need any scheduling and you may guarantee synchronicity by providing a separate sync interface.

What do you think about implementing a synchronous single-threaded subset of zio?

**Comment by jdegoes** [comment]
> Especially in a situation when we cannot save current computation context blocked by an op and serve another op (do "context switch" or "continuation")

This is already done in ZIO, `flatMap` saves a continuation, with a heap-reified stack (of course, it assumes you are "programming in IO" rather than programming outside IO and wrapping large chunks of code in `IO.sync` / `IO.async`). Nonetheless, I am not proposing to do _new_ scheduling as part of this ticket. Only mentioning that having a synchronous, single-threaded `SyncIO` will make it easier to improve scheduling in `IO` proper.

But while I am thinking about it:

 * Another approach to part of these problems is to add some mechanism (like `evalOn`) that causes a fiber to jump to a thread or thread pool; if an `async` constructor is used, it resumes the computation but only back onto the specified thread pool; this adds overhead to `async`, because it only ever immediately submits the continuation to the specified thread pool, but allows fibers to more reliably run on a given thread pool.
 * The main drawback is it doesn't solve the other problems: how do I know that the computation will execute synchronously, on a single thread? I still have no guarantee of this, which is important for some applications.

**Comment by JohnSColeman** [comment]
Sometimes I'd like to execute with RTS on the current thread rather than use the default executor service. Is there a convenient way to do this?

**Comment by jdegoes** [comment]
@JohnSColeman No, although if this ticket is completed, you can execute a `SyncIO` on any thread.


@pshirshov 

> Yup, sure. But what I mean - it would be good if SyncIO not assume any scheduling and executes everything right in the thread which invokes .run. Maybe I misunderstood your idea, but from what I see - you still wish to provide execution context for SyncIO.

I think it should be possible to provide an execution context, or run on the "current" thread. The main point, I think, is that `SyncIO` should never be able to hop threads, which means (a) no yielding, and (b) no async. This gives you a very strong guarantee and an easy way to reason about `SyncIO`.

OTOH, `IO` (AKA "async IO") can hop between threads on any async boundary (it resumes on the thread executing the callback), and even beyond this, it can hop between threads if the async interface is used repeatedly on the same thread (to avoid a stack overflow), or after a certain number of user-defined operations. Reasoning about where `IO` is running is not a simple thing to do; probably you should only use `IO` if you don't care where it's running because the types and (current) model just don't allow you to lock it down to a given thread or thread pool.

Broadly speaking I see two possible (principled) paths forward:

1. Introduce `SyncIO` which has a clear execution model, and encourage people to use `SyncIO` for any case in which it matters what thread (/pool) a fiber is running on. As an added benefit, synchronous IO can have new types of features not possible in async IO, such as blocking IO fiber interruption.
2. Redesign `IO` with a new execution model that's possible to reason about. Practically speaking, I think this would mean that fibers would never jump threads, any use of `IO.async` would simply stuff a new task on the fiber's thread pool (which would mean you cannot force a fiber to execute on a new thread / thread pool, and might be incompatible with some callback APIs), and fibers would never yield.

Of these two options, I have a slight preference for (1) because I feel the vast majority of users don't really care where things execute and enjoy the benefits of relaxed guarantees. But if we could find a better execution model than (2), I might lean toward that instead.

**Comment by kaishh** [comment]
@jdegoes 
> Another approach to part of these problems is to add some mechanism (like evalOn) that causes a fiber to jump to a thread or thread pool; if an async constructor is used, it resumes the computation but only back onto the specified thread pool; this adds overhead to async, because it only ever immediately submits the continuation to the specified thread pool, but allows fibers to more reliably run on a given thread pool.
> The main drawback is it doesn't solve the other problems: how do I know that the computation will execute synchronously, on a single thread? I still have no guarantee of this, which is important for some applications.

These are two separate issues! Some libraries, e.g. http4s want to _both_ spawn a ton of fibers and stay within the same thread pool. IMHO, we have to support `.async`'s staying within the same pool pool purely because of interop with code that assumes that `shift` is enough to ensure execution on a specific pool.

On SyncIO, I think it's a great idea, we should also think if we want `.sync` constructors to go through SyncIO and whether we want to shift all SyncIO's to a separate pool. If we do, should we make `sync` shift by default,  and add a `syncUnsafe` method that doesn't - similar to haskell's safe/unsafe FFI, or instead have a separate `syncBlocking` method?

(Also, on an unrelated note: I think that a non-growing executor with only coreSize threads with no support for `scala.concurrent.blocking` contexts is a bit of an unsafe default choice for the RTS pool, can we change that? Was there a specific motivation for using ExecutorService vs. ExecutionContext?)

**Comment by JohnSColeman** [comment]
@jdegoes I would also like to know why ExecutorService instead of ExecutionContext? We can derive an ES from and EC, but not the other way and it seems counter intuitive for a Scala app. I'd like the RTS to be able to share the EC of the host application for a mixed IO project.
